src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+/// <reference path="../refactorProvider.ts" />
+/// <reference path="../../compiler/checker.ts" />
+
+/* @internal */
+namespace ts.refactor.extractMethod {
+    const extractMethod: Refactor = {
+        name: "Extract Method",
+        description: Diagnostics.Extract_function.message,
+        getAvailableActions,
+        getEditsForAction,
+    };
+
+    registerRefactor(extractMethod);
+
+    /** Compute the associated code actions */
+    function getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+        const rangeToExtract = getRangeToExtract(context.file, { start: context.startPosition, length: context.endPosition - context.startPosition });
+
+        const targetRange: TargetRange = rangeToExtract.targetRange;
+        if (targetRange === undefined) {
+            return undefined;
+        }
+
+        const extractions = extractRange(targetRange, context);
+        if (extractions === undefined) {
+            // No extractions possible
+            return undefined;
+        }
+
+        const actions: RefactorActionInfo[] = [];
+        const usedNames: Map<boolean> = createMap();
+
+        let i = 0;
+        for (const extr of extractions) {
+            // Skip these since we don't have a way to report errors yet
+            if (extr.errors && extr.errors.length) {
+                continue;
+            }
+
+            // Don't issue refactorings with duplicated names

```

Comment:
(leave comment indicating that the inner scope always 'wins' due to the list being order "inner scopes first")